internal/server/boltdbstate/project.go

```diff
@@ -69,10 +69,172 @@ func (s *State) ProjectGet(ref *pb.Ref_Project) (*pb.Project, error) {
 // delete. This will delete all operations associated with this project
 // as well.
 func (s *State) ProjectDelete(ref *pb.Ref_Project) error {
+	// Get our project to delete
+	project, err := s.ProjectGet(ref)
+	if err != nil {
+		if status.Code(err) == codes.NotFound {
+			return nil
+		}
+		return err
+	}
+
+	// We perform all of our reads before our write to avoid the deadlocked state
+	var builds []*pb.Build
+	var artifacts []*pb.PushedArtifact
+	var deployments []*pb.Deployment
+	var releases []*pb.Release
+	var statusReports []*pb.StatusReport
+	var workspaces []*pb.Workspace_Project
+	var triggers []*pb.Trigger
+	var pipelines []*pb.Pipeline
+	var configVars []*pb.ConfigVar
+	if err = s.db.View(func(dbTxn *bolt.Tx) error {
+		for _, app := range project.Applications {
+			appRef := &pb.Ref_Application{
+				Application: app.Name,
+				Project:     project.Name,
+			}
+			if builds, err = s.BuildList(appRef); err != nil {
+				return err
+			}
+			if artifacts, err = s.ArtifactList(appRef); err != nil {
+				return err
+			}
+			if deployments, err = s.DeploymentList(appRef); err != nil {
+				return err
+			}
+			if releases, err = s.ReleaseList(appRef); err != nil {
+				return err
+			}
+			if statusReports, err = s.StatusReportList(appRef); err != nil {
+				return err
+			}
+
+			// Get app-scoped config
+			if appConfigVars, err := s.ConfigGet(&pb.ConfigGetRequest{
+				Scope: &pb.ConfigGetRequest_Application{Application: &pb.Ref_Application{
+					Project:     project.Name,
+					Application: app.Name,
+				}},
+			}); err != nil {
+				return err
+			} else {
+				configVars = append(configVars, appConfigVars...)
+			}
+		}
+
+		// Get project-scoped config
+		if projectConfigVars, err := s.ConfigGet(&pb.ConfigGetRequest{
+			Scope: &pb.ConfigGetRequest_Project{Project: ref},
+		}); err != nil {
+			return err
+		} else {
+			configVars = append(configVars, projectConfigVars...)
+		}
+
+		if workspaceList, err := s.ProjectListWorkspaces(ref); err != nil {
+			return err
+		} else {
+			for _, workspace := range workspaceList {
+				// Get the triggers for a project in the workspace
+				if triggerList, err := s.TriggerList(workspace.Workspace, &pb.Ref_Project{Project: project.Name}, nil, []string{}); err != nil {
+					return err
+				} else {
+					triggers = append(triggers, triggerList...)
+				}
+				if workspaceDetail, err := s.WorkspaceGet(workspace.Workspace.Workspace); err != nil {
+					return err
+				} else {
+					// If the project we're deleting is the only project in the workspace, we delete the workspace
+					// We don't delete the default workspace
+					if len(workspaceDetail.Projects) == 1 && workspace.Workspace.Workspace != "default" {
```

Comment:
Regarding workspaces, one behavior that this PR does not address: 
- there is a workspace with multiple projects
- one of the projects in this workspace is the one we want to delete
- goal: remove the project from the workspace

This isn't yet possible as `func (s *State) WorkspacePut(workspace *pb.Workspace) error` notes:
```go
// WorkspacePut creates or updates the given Workspace.
//
// Project changes will be ignored
```

Prior to learning of this note, locally I was testing getting all the workspaces with multiple projects, removing the project we're deleting from the slice of projects, and then re-putting the workspace. However, until this limitation is addressed, workspaces will continue to display that a deleted project is amongst its projects.

The deletion of workspaces where the project we're deleting is the only project in the workspace is working, though.